packages/lib/src/vis/shared/ZoomSelectionMesh.tsx

```diff
@@ -0,0 +1,62 @@
+import { useThree } from '@react-three/fiber';
+import { useState } from 'react';
+import type { Vector2 } from 'three';
+
+import type { ModifierKey } from '../models';
+import SelectionMesh from './SelectionMesh';
+import SelectionRect from './SelectionRect';
+import { useMoveCameraTo } from './hooks';
+
+interface Props {
+  modifierKey?: ModifierKey;
+}
+
+function ZoomSelectionMesh(props: Props) {
+  const { modifierKey } = props;
+
+  const [startPoint, setStartPoint] = useState<Vector2>();
+
+  const moveCameraTo = useMoveCameraTo();
+
+  const { width, height } = useThree((state) => state.size);
+  const camera = useThree((state) => state.camera);
+
+  const onSelectionEnd = (endPoint: Vector2) => {
+    if (!startPoint) {
+      return;
+    }
+    if (startPoint.x === endPoint.x && startPoint.y === endPoint.y) {
+      return;
+    }
+
+    const selectionWidth = Math.abs(endPoint.x - startPoint.x);
+    const selectionHeight = Math.abs(endPoint.y - startPoint.y);
+
+    // Change scale first so that moveCameraTo computes the updated camera bounds
+    camera.scale.set(selectionWidth / width, selectionHeight / height, 1);
+    camera.updateProjectionMatrix();
+    camera.updateMatrixWorld();
+
+    const selectionCenter = endPoint
+      .clone()
+      .sub(startPoint)
+      .divideScalar(2)
+      .add(startPoint);
+
+    moveCameraTo(selectionCenter.x, selectionCenter.y);
+  };
+
+  return (
+    <SelectionMesh
+      modifierKey={modifierKey}
+      onSelectionStart={setStartPoint}
+      onSelectionEnd={onSelectionEnd}
```

Comment:
Following on my previous two comments in the stories, I don't think this component needs a state at all. I think the callbacks should be declared like this (with `onSelectionStart` removed):

```ts
onSelectionEnd={handleSelectionEnd}
```

with `function handleSelectionEnd(selection:Selection): void { ... }`